### PR TITLE
Migrate xcaddy to Cobra

### PR DESCRIPTION
**Pull request by armadi1809**
This is still in draft mode (early stage) and not ready to be merged yet. Putting it here to get initial feedback (if possible) as I continue to work through this. I am also still finalizing the details about the proposed architecture.

### CODE REVIEW

**Comment by mholt** [comment]
Neat, I think this is looking good so far. I don't know if maybe @mohammed90 wants to give some additional feedback or his thoughts when he has time.

**Comment by armadi1809** [comment]
You reminded me of this lol 😅  I still need to work on some last details. I will post here again once that's done.

**Comment by armadi1809** [comment]
I think this is in a good spot for an initial review and perhaps some followup discussions.

**Comment by mholt** [comment]
Sorry I lost track of this @armadi1809. Does this change the command line interface of xcaddy at all, or is it identical?

File: cmd/commands.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	"runtime"
+	"strings"
+
+	"github.com/caddyserver/xcaddy"
+	"github.com/caddyserver/xcaddy/internal/utils"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	buildCommand.Flags().String("with", "", "can be used multiple times to add plugins by specifying the Go module name and optionally its version, similar to go get. Module name is required, but specific version and/or local replacement are optional.")
```
Comment (mohammed90) [inline cmd/commands.go:18]: The `--with` flag can be specified multiple times, so this should be a string array

File: cmd/commands.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	buildCommand.Flags().String("output", "", "changes the output file.")
+	buildCommand.Flags().String("replace", "", "is like --with, but does not add a blank import to the code; it only writes a replace directive to go.mod, which is useful when developing on Caddy's dependencies (ones that are not Caddy modules). Try this if you got an error when using --with, like \"cannot find module providing package\".")
```
Comment (mohammed90) [inline cmd/commands.go:20]: Similar to `--with`, the `--replace` can be specified multiple times

File: cmd/commands.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	buildCommand.Flags().StringSlice("embed", []string{}, "can be used multiple times to embed directories into the built Caddy executable. The directory can be prefixed with a custom alias and a colon : to use it with the root directive and sub-directive.")
```
Comment (mohammed90) [inline cmd/commands.go:21]: Use `StringArray` rather then slice because we don't accept comma-separated values for this flag

**Review by mohammed90 — CHANGES_REQUESTED** [review]
Thank you for tackling this! As with the others, I'll most likely do the review in rounds 😅 I like where this is going. Here's the first round of review.

**Comment by armadi1809** [comment]
Thanks for this first round of reviews Mohammed. All the issues should now be resolved.

**Comment by pcfreak30** [comment]
@mohammed90 Just wanted to say thanks for this effort. ive actually forked xcaddy for my own project to use the same build process (it is inspired by caddy's architecture) and this PR will also enable me to adopt the cli library.

File: cmd/cobra.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	"os/exec"
+
+	"github.com/caddyserver/xcaddy"
+	"github.com/caddyserver/xcaddy/internal/utils"
+	"github.com/spf13/cobra"
+)
+
+var rootCmd = &cobra.Command{
+	Use:          "xcaddy",
+	Long:         "",
```
Comment (mohammed90) [inline cmd/cobra.go:16]: Please add content for the `Short` and `Long` fields

File: cmd/commands.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+func init() {
+	buildCommand.Flags().StringArray("with", []string{}, "can be used multiple times to add plugins by specifying the Go module name and optionally its version, similar to go get. Module name is required, but specific version and/or local replacement are optional.")
+	buildCommand.Flags().String("output", "", "changes the output file.")
+	buildCommand.Flags().StringArray("replace", []string{}, "is like --with, but does not add a blank import to the code; it only writes a replace directive to go.mod, which is useful when developing on Caddy's dependencies (ones that are not Caddy modules). Try this if you got an error when using --with, like \"cannot find module providing package\".")
+	buildCommand.Flags().StringArray("embed", []string{}, "can be used multiple times to embed directories into the built Caddy executable. The directory can be prefixed with a custom alias and a colon : to use it with the root directive and sub-directive.")
+}
+
+var versionCommand = &cobra.Command{
+	Use:  "version",
+	Long: "Getting xcaddy version",
```
Comment (mohammed90) [inline cmd/commands.go:26]: Same here -- add `Short` and `Long`

**Review by mohammed90 — CHANGES_REQUESTED** [review]
We're getting closer, but there's a use-case that's broken now. It's possible to run `xcaddy validate`, `xcaddy list-modules`, or any caddy command -- which xcaddy takes and passes to caddy as-is. This is broken now. This is what I get:
```
/workspaces/xcaddy/caddy-git-fs (main) $ ../cmd/xcaddy/xcaddy list-modules
Error: unknown command "list-modules" for "xcaddy"
Run 'xcaddy --help' for usage.
unknown command "list-modules" for "xcaddy"
```

We need to figure it out because it's critical part of the Caddy module development.

File: cmd/cobra.go (hunk 4 above)
Comment (armadi1809) [reply cmd/cobra.go]: I added short and long fields but still want your opinion on what should go/not go in these.

File: cmd/commands.go (hunk 5 above)
Comment (armadi1809) [reply cmd/commands.go]: Same as other comment

**Comment by armadi1809** [comment]
@mohammed90 The issue with the arguments to be passed to caddy should be resolved. I still want your opinion on the commands descriptions (short and long) though as I am not entirely sure about what exactly should go in these. Thanks.

File: cmd/commands.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	"github.com/caddyserver/xcaddy"
+	"github.com/caddyserver/xcaddy/internal/utils"
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	buildCommand.Flags().StringArray("with", []string{}, "can be used multiple times to add plugins by specifying the Go module name and optionally its version, similar to go get. Module name is required, but specific version and/or local replacement are optional.")
+	buildCommand.Flags().String("output", "", "changes the output file.")
+	buildCommand.Flags().StringArray("replace", []string{}, "is like --with, but does not add a blank import to the code; it only writes a replace directive to go.mod, which is useful when developing on Caddy's dependencies (ones that are not Caddy modules). Try this if you got an error when using --with, like \"cannot find module providing package\".")
+	buildCommand.Flags().StringArray("embed", []string{}, "can be used multiple times to embed directories into the built Caddy executable. The directory can be prefixed with a custom alias and a colon : to use it with the root directive and sub-directive.")
```
Comment (mohammed90) [inline cmd/commands.go:21]: Let's make the description shorter here but have the longer form in the `Long` section of the command.

```suggestion
	buildCommand.Flags().StringArray("with", []string{}, "caddy modules package path to include in the build")
	buildCommand.Flags().String("output", "", "change the output file name")
	buildCommand.Flags().StringArray("replace", []string{}, "like --with but for Go modules")
	buildCommand.Flags().StringArray("embed", []string{}, "embeds directories into the built Caddy executable to use with the `embedded` file-system")
```

File: cmd/commands.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+var versionCommand = &cobra.Command{
+	Use:   "version",
+	Long:  "Command to get xcaddy version",
+	Short: "xcaddy version",
```
Comment (mohammed90) [inline cmd/commands.go:27]: We may be able to eliminate one of them. I'm not sure what will the display be like when each is eliminated.

```suggestion
	Long:  "Prints xcaddy version",
	Short: "Prints xcaddy version",
```

**Review by mohammed90 — CHANGES_REQUESTED** [review]
I've amended the help lines, but I'm not very happy with my phrasings either. If @mholt and @francislavoie can take a look, that'd be great. Y'all have a good eye for this.

File: cmd/cobra.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	Use: "xcaddy <args...>",
+	Long: `If you run xcaddy from within the folder of the Caddy plugin you're working on without the build subcommand, it will build Caddy with your current module and run it, as if you manually plugged it in and invoked go run.
+
+The binary will be built and run from the current directory, then cleaned up.
+
+The current working directory must be inside an initialized Go module.
+Syntax:
+ xcaddy <args...>
+ * <args...> are passed through to the caddy command.`,
+	Short:        "A replacement for go run while developing Caddy plugins",
```
[the remark this answers is not in the document]
Comment (mholt) [reply cmd/cobra.go:24]: Ah good. I was going to suggest basically the same thing!

File: cmd/commands.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+A commit like a58f240d3ecbb59285303746406cab50217f8d24
+
+Flags: 
+ --output changes the output file.
+
+ --with can be used multiple times to add plugins by specifying the Go module name and optionally its version, similar to go get. Module name is required, but specific version and/or local replacement are optional.
+
+ --replace is like --with, but does not add a blank import to the code; it only writes a replace directive to go.mod, which is useful when developing on Caddy's dependencies (ones that are not Caddy modules). Try this if you got an error when using --with, like cannot find module providing package.
+
+ --embed can be used multiple times to embed directories into the built Caddy executable. The directory can be prefixed with a custom alias and a colon : to use it with the root directive and sub-directive.
```
Comment (mholt) [inline cmd/commands.go:55]: Do you think the alias thing could use an example?

**Review by mholt — COMMENTED** [review]
tbh I think @mohammed90 's suggestions are great, and I would have done about the same.

Thanks for this change, it's looking good!

File: cmd/commands.go (hunk 9 above)
Comment (armadi1809) [reply cmd/commands.go]: Added an example for the alias.

File: cmd/commands.go (hunk 7 above)
Comment (armadi1809) [reply cmd/commands.go]: Kept short as it's the one that gets displayed for both help commands (at the tool and the command levels)

File: cmd/commands.go (hunk 6 above)
Comment (armadi1809) [reply cmd/commands.go]: Applied this suggestion. Thank you!

File: cmd/cobra.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (armadi1809) [reply cmd/cobra.go]: Applied this suggestion.

**Review by mholt — APPROVED** [review]
Maybe we can give this a shot. What do you think, @mohammed90 ?

**Review by mohammed90 — APPROVED** [review]
Thank you, @armadi1809! LGTM. Let's go!

File: cmd/commands.go [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+
+		replaceArgs, err := cmd.Flags().GetStringArray("replace")
+		if err != nil {
+			return fmt.Errorf("unable to parse --replace arguments: %s", err.Error())
+		}
+		for _, withArg := range withArgs {
+			mod, ver, _, err := splitWith(withArg)
+			if err != nil {
+				return err
+			}
```
Comment (mohammed90) [inline cmd/commands.go:80]: @armadi1809 The third returned value here, which was called `repl` in earlier [code](https://github.com/caddyserver/xcaddy/blob/ff67f855c7a41a07d95849d90372aee939456fb4/cmd/main.go#L83-L108), is ignored. It shouldn't be ignored. It shouldn't be.

Can you copy the older logic for the fix? If you're too busy, I can do it.

File: cmd/commands.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+			// adjust relative replacements in current working directory since our temporary module is in a different directory
+			if strings.HasPrefix(repl, ".") {
+				repl, err = filepath.Abs(repl)
+				if err != nil {
+					log.Fatalf("[FATAL] %v", err)
+				}
+				log.Printf("[INFO] Resolved relative replacement %s to %s", withArg, repl)
+			}
+			replacements = append(replacements, xcaddy.NewReplace(xcaddy.Dependency{PackagePath: mod, Version: ver}.String(), repl))
```
Comment (mohammed90) [inline cmd/commands.go:101]: Continuing the comment [here](https://github.com/caddyserver/xcaddy/pull/174/files#r1753778241), Basically, these lines are applicable to both `--with` and `--replace`.

**Comment by armadi1809** [comment]
Thanks for taking care of this @mohammed90 !

**Pull request merged**